File: attached_assets/extracted/develop/app/services/decision_engine.py
from __future__ import annotations

import math
from datetime import datetime
from typing import Optional, List

from app.models.schemas import (
    PriceForecast,
    SolarForecastSeries,
    BatteryState,
    PlanStep,
    DecisionOutcome,
)


class DecisionEngine:
    """
    Deterministic, explainable planner that weighs hourly solar and grid prices,
    EV/home battery constraints, and a target deadline.
    """

    def __init__(self, feed_in_tariff: float, export_allowed: bool = True):
        self.feed_in_tariff = feed_in_tariff
        self.export_allowed = export_allowed

    def _future_prices(self, prices: List[float], start: int) -> List[float]:
        return prices[start:] if start < len(prices) else []

    def plan(
        self,
        now: datetime,
        price_forecast: PriceForecast,
        solar_forecast: SolarForecastSeries,
        ev_battery: BatteryState,
        target_soc: float,
        deadline_hours: int,
        home_battery: Optional[BatteryState] = None,
    ) -> DecisionOutcome:
        price_points = price_forecast.points
        solar_points = solar_forecast.points
        horizon = min(deadline_hours, len(price_points))

        ev_target_kwh = ev_battery.capacity_kwh * target_soc
        ev_needed = max(0.0, ev_target_kwh - ev_battery.soc_kwh)
        ev_start_needed = ev_needed

        net_cost = 0.0
        total_grid = 0.0
        total_solar_used = 0.0
        total_export = 0.0
        steps: List[PlanStep] = []

        price_list = [p.price_per_kwh for p in price_points]
        avg_grid_price = sum(price_list[:horizon]) / horizon if horizon else 0.0
        baseline_all_grid = ev_needed * avg_grid_price

        for h in range(horizon):
            price = price_points[min(h, len(price_points) - 1)].price_per_kwh
            solar = solar_points[min(h, len(solar_points) - 1)].energy_kwh
            action_notes: List[str] = []

            # 1) Use solar for EV within charge limit
            solar_to_ev = min(
                solar, ev_needed, ev_battery.max_charge_kw
            )
            if solar_to_ev > 0:
                ev_battery.soc_kwh += solar_to_ev
                ev_needed -= solar_to_ev
                solar -= solar_to_ev
                total_solar_used += solar_to_ev
                action_notes.append(f"Used {solar_to_ev:.2f} kWh solar for EV")

            # 2) Store solar into home battery if future prices are higher
            home_to_store = 0.0
            if home_battery and solar > 0:
                future_prices = self._future_prices(price_list, h)
                future_max = max(future_prices) if future_prices else price
                if future_max > price + 0.05:
                    capacity_left = home_battery.capacity_kwh - home_battery.soc_kwh
                    if capacity_left > 0:
                        home_to_store = min(
                            solar, capacity_left, home_battery.max_charge_kw
                        )
                        home_battery.soc_kwh += home_to_store
                        solar -= home_to_store
                        action_notes.append(
                            f"Stored {home_to_store:.2f} kWh solar in home battery for later high prices"
                        )

            # 3) Export remaining solar
            exported = 0.0
            if self.export_allowed and solar > 0:
                exported = solar
                net_cost -= exported * self.feed_in_tariff
                total_export += exported
                action_notes.append(
                    f"Exported {exported:.2f} kWh at ${self.feed_in_tariff:.2f}/kWh"
                )
                solar = 0.0

            # 4) Use home battery to avoid expensive grid during peaks
            home_to_ev = 0.0
            if home_battery and ev_needed > 0 and home_battery.soc_kwh > 0:
                future_prices = self._future_prices(price_list, h)
                cheapest_future = min(future_prices) if future_prices else price
                if price >= cheapest_future + 0.08 or price >= 0.30:
                    home_to_ev = min(
                        ev_needed, home_battery.soc_kwh, home_battery.max_discharge_kw
                    )
                    home_battery.soc_kwh -= home_to_ev
                    ev_needed -= home_to_ev
                    total_solar_used += home_to_ev
                    action_notes.append(
                        f"Shifted {home_to_ev:.2f} kWh from home battery to EV to avoid high price"
                    )

            # 5) Decide grid charging for this hour
            grid_to_ev = 0.0
            if ev_needed > 0:
                hours_left = max(0, deadline_hours - h - 1)
                required_hours_min = (
                    math.ceil(ev_needed / ev_battery.max_charge_kw)
                    if ev_battery.max_charge_kw > 0
                    else 0
                )
                future_prices = self._future_prices(price_list, h)
                cheapest_future = min(future_prices) if future_prices else price
                should_charge_now = price <= cheapest_future + 0.01 or hours_left < required_hours_min

                if should_charge_now:
                    grid_to_ev = min(ev_needed, ev_battery.max_charge_kw)
                    ev_battery.soc_kwh += grid_to_ev
                    ev_needed -= grid_to_ev
                    cost = grid_to_ev * price
                    net_cost += cost
                    total_grid += grid_to_ev
                    action_notes.append(
                        f"Bought {grid_to_ev:.2f} kWh from grid at ${price:.2f}/kWh"
                    )
                else:
                    action_notes.append("Deferred grid charging awaiting cheaper window")

            steps.append(
                PlanStep(
                    hour=h,
                    grid_price=price,
                    solar_used_kwh=round(solar_to_ev, 3),
                    grid_used_kwh=round(grid_to_ev, 3),
                    home_used_kwh=round(home_to_ev, 3),
                    exported_kwh=round(exported, 3),
                    ev_soc_kwh=round(ev_battery.soc_kwh, 3),
                    home_soc_kwh=round(home_battery.soc_kwh, 3) if home_battery else None,
                    action=", ".join(action_notes) if action_notes else "No action",
                    note=f"Price window: ${price:.2f}/kWh",
                )
            )

            if ev_needed <= 0:
                break

        recommendation = "wait" if ev_needed > 0 else "charge_optimized"
        if ev_needed > 0:
            steps.append(
                PlanStep(
                    hour=horizon,
                    grid_price=price_list[min(horizon - 1, len(price_list) - 1)],
                    solar_used_kwh=0,
                    grid_used_kwh=0,
                    home_used_kwh=0,
                    exported_kwh=0,
                    ev_soc_kwh=round(ev_battery.soc_kwh, 3),
                    home_soc_kwh=round(home_battery.soc_kwh, 3) if home_battery else None,
                    action="Incomplete target",
                    note=f"Still need {ev_needed:.2f} kWh by deadline",
                )
            )

        savings = baseline_all_grid - net_cost
        explanation = (
            f"Target: {ev_target_kwh:.2f} kWh by +{deadline_hours}h. "
            f"Used {total_solar_used:.2f} kWh solar, {total_grid:.2f} kWh grid. "
            f"Exported {total_export:.2f} kWh. "
            f"Savings vs all-grid plan: ${savings:.2f}."
        )

        return DecisionOutcome(
            recommendation=recommendation,
            net_cost=round(net_cost, 2),
            total_grid_kwh=round(total_grid, 2),
            total_solar_used_kwh=round(total_solar_used, 2),
            total_export_kwh=round(total_export, 2),
            savings_vs_all_grid=round(savings, 2),
            steps=steps,
            explanation=explanation,
        )

File: server/services/decision_engine.ts
import { 
  PriceForecast, 
  SolarForecastSeries, 
  BatteryState, 
  PlanStep, 
  DecisionOutcome, 
  PricePoint,
  SolarPoint
} from "@shared/schema";

export class DecisionEngine {
  private feedInTariff: number;
  private exportAllowed: boolean;

  constructor(feedInTariff: number, exportAllowed: boolean = true) {
    this.feedInTariff = feedInTariff;
    this.exportAllowed = exportAllowed;
  }

  private futurePrices(prices: number[], startExclusive: number): number[] {
    const idx = startExclusive + 1;
    return idx < prices.length ? prices.slice(idx) : [];
  }

  public plan(
    now: Date,
    priceForecast: PriceForecast,
    solarForecast: SolarForecastSeries,
    evBattery: BatteryState,
    targetSoc: number,
    deadlineHours: number,
    homeBattery?: BatteryState
  ): DecisionOutcome {
    const pricePoints = priceForecast.points;
    const solarPoints = solarForecast.points;
    const horizon = Math.min(deadlineHours, pricePoints.length);

    const evTargetKwh = evBattery.capacity_kwh * targetSoc;
    let evNeeded = Math.max(0.0, evTargetKwh - evBattery.soc_kwh);
    const evStartNeeded = evNeeded;

    let netCost = 0.0;
    let totalGrid = 0.0;
    let totalSolarUsed = 0.0;
    let totalExport = 0.0;
    const steps: PlanStep[] = [];

    const priceList = pricePoints.map(p => p.price_per_kwh);
    const avgGridPrice = horizon > 0 
      ? priceList.slice(0, horizon).reduce((a, b) => a + b, 0) / horizon 
      : 0.0;
    const baselineAllGrid = evNeeded * avgGridPrice;

    // We modify local copies of battery state to track simulation
    const currentEvSoc = { ...evBattery };
    const currentHomeSoc = homeBattery ? { ...homeBattery } : undefined;

    // --- Pre-plan grid charging on the cheapest hours (Fix #2) ---
    const solarUsable = solarPoints.slice(0, horizon).map(p =>
      Math.min(p.energy_kwh, currentEvSoc.max_charge_kw)
    );
    const expectedSolar = solarUsable.reduce((a, b) => a + b, 0);
    let remainingForGridPlan = Math.max(0, evNeeded - expectedSolar);

    const hourOrderByPrice = [...Array(horizon).keys()].sort(
      (a, b) => priceList[a] - priceList[b]
    );
    const plannedGrid: number[] = Array(horizon).fill(0);
    for (const h of hourOrderByPrice) {
      if (remainingForGridPlan <= 0) break;
      const alloc = Math.min(remainingForGridPlan, currentEvSoc.max_charge_kw);
      plannedGrid[h] = alloc;
      remainingForGridPlan -= alloc;
    }

    for (let h = 0; h < horizon; h++) {
      const priceIndex = Math.min(h, pricePoints.length - 1);
      const solarIndex = Math.min(h, solarPoints.length - 1);
      
      const price = pricePoints[priceIndex].price_per_kwh;
      let solar = solarPoints[solarIndex].energy_kwh;
      const actionNotes: string[] = [];

      // 1) Use solar for EV within charge limit
      const solarToEv = Math.min(
        solar, 
        evNeeded, 
        currentEvSoc.max_charge_kw
      );

      if (solarToEv > 0) {
        currentEvSoc.soc_kwh += solarToEv;
        evNeeded -= solarToEv;
        solar -= solarToEv;
        totalSolarUsed += solarToEv;
        actionNotes.push(`Used ${solarToEv.toFixed(2)} kWh solar for EV`);
      }

      // 2) Store solar into home battery if future prices are higher
      let homeToStore = 0.0;
      if (currentHomeSoc && solar > 0) {
        const futurePricesList = this.futurePrices(priceList, h);
        const futureMax = futurePricesList.length > 0 ? Math.max(...futurePricesList) : price;
        
        if (futureMax > price + 0.05) {
          const capacityLeft = currentHomeSoc.capacity_kwh - currentHomeSoc.soc_kwh;
          if (capacityLeft > 0) {
            homeToStore = Math.min(
              solar,
              capacityLeft,
              currentHomeSoc.max_charge_kw
            );
            currentHomeSoc.soc_kwh += homeToStore;
            solar -= homeToStore;
            actionNotes.push(`Stored ${homeToStore.toFixed(2)} kWh solar in home battery`);
          }
        }
      }

      // 3) Export remaining solar
      let exported = 0.0;
      if (this.exportAllowed && solar > 0) {
        exported = solar;
        netCost -= exported * this.feedInTariff;
        totalExport += exported;
        actionNotes.push(`Exported ${exported.toFixed(2)} kWh at $${this.feedInTariff.toFixed(2)}/kWh`);
        solar = 0.0;
      }

      // 4) Use home battery to avoid expensive grid during peaks
      let homeToEv = 0.0;
      if (currentHomeSoc && evNeeded > 0 && currentHomeSoc.soc_kwh > 0) {
        const futurePricesList = this.futurePrices(priceList, h);
        const cheapestFuture = futurePricesList.length > 0 ? Math.min(...futurePricesList) : price;
        
        if (price >= cheapestFuture + 0.08 || price >= 0.30) {
          homeToEv = Math.min(
            evNeeded,
            currentHomeSoc.soc_kwh,
            currentHomeSoc.max_discharge_kw
          );
          currentHomeSoc.soc_kwh -= homeToEv;
          evNeeded -= homeToEv;
          totalSolarUsed += homeToEv; // Counting stored solar as solar used
          actionNotes.push(`Shifted ${homeToEv.toFixed(2)} kWh from home battery to EV`);
        }
      }

      // 5) Decide grid charging for this hour
      let gridToEv = 0.0;
      if (evNeeded > 0) {
        const hoursLeft = Math.max(0, deadlineHours - h - 1);
        const requiredHoursMin = currentEvSoc.max_charge_kw > 0 
          ? Math.ceil(evNeeded / currentEvSoc.max_charge_kw) 
          : 0;
        
        const futurePricesList = this.futurePrices(priceList, h);
        const cheapestFuture = futurePricesList.length > 0 ? Math.min(...futurePricesList) : price;
        const planned = plannedGrid[h] || 0;
        const shouldChargeNow = planned > 0 || hoursLeft < requiredHoursMin;

        if (shouldChargeNow) {
          const plannedAmount = Math.min(planned, evNeeded, currentEvSoc.max_charge_kw);
          const emergencyAmount =
            hoursLeft < requiredHoursMin ? Math.min(evNeeded, currentEvSoc.max_charge_kw) : 0;
          gridToEv = Math.max(plannedAmount, emergencyAmount);

          currentEvSoc.soc_kwh += gridToEv;
          evNeeded -= gridToEv;
          const cost = gridToEv * price;
          netCost += cost;
          totalGrid += gridToEv;
          actionNotes.push(`Bought ${gridToEv.toFixed(2)} kWh from grid at $${price.toFixed(2)}/kWh`);
        } else {
          actionNotes.push("Deferred grid charging awaiting cheaper window");
        }
      }

      steps.push({
        hour: h,
        grid_price: price,
        solar_used_kwh: Number(solarToEv.toFixed(3)),
        grid_used_kwh: Number(gridToEv.toFixed(3)),
        home_used_kwh: Number(homeToEv.toFixed(3)),
        exported_kwh: Number(exported.toFixed(3)),
        ev_soc_kwh: Number(currentEvSoc.soc_kwh.toFixed(3)),
        home_soc_kwh: currentHomeSoc ? Number(currentHomeSoc.soc_kwh.toFixed(3)) : null,
        action: actionNotes.length > 0 ? actionNotes.join(", ") : "No action",
        note: `Price window: $${price.toFixed(2)}/kWh`
      });

      if (evNeeded <= 0) break; // Fix #1: stop immediately once target met
    }

    const recommendation = evNeeded > 0 ? "wait" : "charge_optimized";
    
    // Add final step if incomplete
    if (evNeeded > 0) {
        const lastPrice = priceList[Math.min(horizon - 1, priceList.length - 1)];
        steps.push({
            hour: horizon,
            grid_price: lastPrice,
            solar_used_kwh: 0,
            grid_used_kwh: 0,
            home_used_kwh: 0,
            exported_kwh: 0,
            ev_soc_kwh: Number(currentEvSoc.soc_kwh.toFixed(3)),
            home_soc_kwh: currentHomeSoc ? Number(currentHomeSoc.soc_kwh.toFixed(3)) : null,
            action: "Incomplete target",
            note: `Still need ${evNeeded.toFixed(2)} kWh by deadline`
        });
    }

    const savings = baselineAllGrid - netCost;
    const explanation = `Target: ${evTargetKwh.toFixed(2)} kWh by +${deadlineHours}h. ` +
      `Used ${totalSolarUsed.toFixed(2)} kWh solar, ${totalGrid.toFixed(2)} kWh grid. ` +
      `Exported ${totalExport.toFixed(2)} kWh. ` +
      `Savings vs all-grid plan: $${savings.toFixed(2)}.`;

    return {
      recommendation,
      net_cost: Number(netCost.toFixed(2)),
      total_grid_kwh: Number(totalGrid.toFixed(2)),
      total_solar_used_kwh: Number(totalSolarUsed.toFixed(2)),
      total_export_kwh: Number(totalExport.toFixed(2)),
      savings_vs_all_grid: Number(savings.toFixed(2)),
      steps,
      explanation
    };
  }
}
